metagraph/tests/test_vector_algorithm.cpp

```diff
@@ -160,106 +160,66 @@ TEST(IntVector, call_nonzeros_sparse_every_4) {
     }
 }
 
-TEST(IntVector, atomic_exchange_and_fetch_sparse) {
+TEST(IntVector, atomic_exchange_and_fetch) {
     for (size_t w = 1; w <= 64; ++w) {
         for (auto memorder : { __ATOMIC_RELAXED, __ATOMIC_SEQ_CST }) {
-            sdsl::int_vector<> vector_atomic = aligned_int_vector(200, 0, w, 16);
-            sdsl::int_vector<> vector(200, 0, w);
-            ASSERT_LT(65u, vector.size());
-            std::atomic<size_t> counter = 0;
+            sdsl::int_vector<> vector_atomic = aligned_int_vector(600, 0, w, 16);
+            uint64_t val = sdsl::bits::lo_set[w];
 
             std::mutex mu;
-            std::mutex mu2;
+            std::atomic_thread_fence(std::memory_order_release);
             #pragma omp parallel for num_threads(3) schedule(static, 1)
-            for (size_t i = 0; i < vector.size(); ++i) {
-                if (i % 130 == 1) {
-                    uint64_t val = (++counter) % w;
-                    EXPECT_EQ(0u, atomic_exchange(vector_atomic, i, val, mu, memorder));
-                    std::lock_guard<std::mutex> lock(mu2);
-                    vector[i] = val;
-                }
+            for (size_t i = 0; i < vector_atomic.size(); ++i) {
+                EXPECT_EQ(0u, atomic_exchange(vector_atomic, i, val, mu, memorder));
             }
 
-            for (size_t i = 0; i < vector.size(); ++i) {
-                EXPECT_EQ(vector[i], atomic_fetch(vector_atomic, i, mu, __ATOMIC_ACQUIRE));
-            }
+            std::atomic_thread_fence(std::memory_order_acquire);
+            EXPECT_EQ(sdsl::int_vector<>(600, val, w), vector_atomic);
         }
     }
 }
 
-TEST(IntVector, atomic_fetch_and_add_sparse) {
+TEST(IntVector, atomic_fetch_and_add_val) {
     for (size_t w = 1; w <= 64; ++w) {
         for (auto memorder : { __ATOMIC_RELAXED, __ATOMIC_SEQ_CST }) {
-            sdsl::int_vector<> vector_atomic = aligned_int_vector(200, 0, w, 16);
-            sdsl::int_vector<> vector(200, 0, w);
-            ASSERT_LT(65u, vector.size());
+            sdsl::int_vector<> vector_atomic = aligned_int_vector(600, 0, w, 16);
+            uint64_t val = std::min(uint64_t(500), sdsl::bits::lo_set[w]);
 
             std::mutex mu;
-            std::mutex mu2;
+            std::atomic_thread_fence(std::memory_order_release);
             #pragma omp parallel for num_threads(3) schedule(static, 1)
-            for (size_t i = 0; i < vector.size(); ++i) {
-                if (i % 130 == 1) {
-                    EXPECT_EQ(0u, atomic_fetch_and_add(vector_atomic, i, 1, mu, memorder));
-                    std::lock_guard<std::mutex> lock(mu2);
-                    ++vector[i];
+            for (size_t j = 0; j < val; ++j) {
+                for (size_t i = 0; i < vector_atomic.size(); ++i) {
+                    atomic_fetch_and_add(vector_atomic, i, 1, mu, memorder);
```

Comment:
maybe also try updating only a single element of the vector `atomic_fetch_and_add(vector_atomic, 0, ...)` to increase the contention. Otherwise, it's quite unlikely that two threads are updating exactly the same element at the same time